streaming/src/main/scala/io/gearpump/streaming/dsl/StreamApp.scala

```diff
@@ -78,15 +78,35 @@ object StreamApp {
   }
 }
 
-trait TypedDataSource[T] extends DataSource
+object TimeExtractor {
+  def apply[T](fn: T => TimeStamp): TimeExtractor[T] = new TimeExtractor[T] {
+    override def apply(t: T): TimeStamp = fn(t)
+  }
+}
+trait TimeExtractor[T] {
+  def apply(t: T): TimeStamp
+}
+
+trait TypedDataSource[T] extends DataSource {
+  def setTimeExtractor(extractor: TimeExtractor[T]): Unit
+}
 
 class CollectionDataSource[T](seq: Seq[T]) extends TypedDataSource[T] {
   var index = 0
+  var timeExtractor: TimeExtractor[T] = TimeExtractor((t: T) => System.currentTimeMillis())
+
+  override def setTimeExtractor(extractor: TimeExtractor[T]): Unit = {
+    this.timeExtractor = extractor
+  }
 
   override def read(batchSize: Int): List[Message] = {
-    val size = Math.min(seq.length - index % seq.length, batchSize)
-    index += size
-    seq.slice(index % seq.length - size, size).toList.map(msg => Message(msg.asInstanceOf[AnyRef], System.currentTimeMillis()))
+    if (seq.isEmpty) {
+      List.empty[Message]
+    } else {
+      val size = Math.min(seq.length - index % seq.length, batchSize)
+      index += size
+      seq.slice(index % seq.length - size, size).toList.map(msg => Message(msg.asInstanceOf[AnyRef], timeExtractor(msg)))
```

Comment:
Can the calculation of `startIndex` and `stopIndex` of `slice` be more straightforward?